File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Solution {
	
	static Deque<Character> queue;
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		for (int test_case = 1; test_case<11; test_case++) {
			int n = Integer.parseInt(br.readLine());
			char[] data = br.readLine().toCharArray();
			queue = new ArrayDeque<>();
			
			for (int i = 0; i<n; i++) {
				if (data[i] == '(') {
					queue.offer(data[i]);
				}
				if (data[i] == '[') {
					queue.offer(data[i]);
				}
				if (data[i] == '{') {
					queue.offer(data[i]);
				}
				if (data[i] == '<') {
					queue.offer(data[i]);
				}
				if (data[i] == '}') {
					if (queue.size() == 0) {
						queue.offer(data[i]);
					}
					else {
						if (queue.peekLast() == '{') {
							queue.pollLast();
						}
						else {
							queue.offer(data[i]);
						}
					}
				}
				if (data[i] == '>') {
					if (queue.size() == 0) {
						queue.offer(data[i]);
					}
					else {
						if (queue.peekLast() == '<') {
							queue.pollLast();
						}
						else {
							queue.offer(data[i]);
						}
					}
				}
				if (data[i] == ']') {
					if (queue.size() == 0) {
						queue.offer(data[i]);
					}
					else {
						if (queue.peekLast() == '[') {
							queue.pollLast();
						}
						else {
							queue.offer(data[i]);
						}
					}
				}
				if (data[i] == ')') {
					if (queue.size() == 0) {
						queue.offer(data[i]);
					}
					else {
						if (queue.peekLast() == '(') {
							queue.pollLast();
						}
						else {
							queue.offer(data[i]);
						}
					}
				}
				
			}
			if (queue.size() == 0) {
				System.out.println("#"+test_case+" "+1);
			}
			else {
				System.out.println("#"+test_case+" "+0);
			}
		}
	}
}

File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.py
from collections import deque

for test_case in range(1, 11):
    n = int(input().rstrip())
    data = input().rstrip()
    queue = deque()
    for i in data:
        if i == "(":
            queue.append(i)
        if i == "[":
            queue.append(i)
        if i == "{":
            queue.append(i)
        if i == "<":
            queue.append(i)
        if i == "}":
            if len(queue) == 0:
                queue.append(i)
            else:
                if queue[-1] == "{":
                    queue.pop()
                else:
                    queue.append(i)
        if i == "]":
            if len(queue) == 0:
                queue.append(i)
            else:
                if queue[-1] == "[":
                    queue.pop()
                else:
                    queue.append(i)
        if i == ")":
            if len(queue) == 0:
                queue.append(i)
            else:
                if queue[-1] == "(":
                    queue.pop()
                else:
                    queue.append(i)
        if i == ">":
            if len(queue) == 0:
                queue.append(i)
            else:
                if queue[-1] == "<":
                    queue.pop()
                else:
                    queue.append(i)
    if len(queue) == 0:
        print("#" + str(test_case) + " " + str(1))
    else:
        print("#" + str(test_case) + " " + str(0))